src/main/java/iroha/validation/transactions/storage/impl/BasicTransactionProvider.java

```diff
@@ -0,0 +1,106 @@
+package iroha.validation.transactions.storage.impl;
+
+import com.google.common.base.Strings;
+import io.reactivex.Observable;
+import iroha.protocol.BlockOuterClass.Block;
+import iroha.protocol.Queries;
+import iroha.protocol.Queries.BlocksQuery;
+import iroha.protocol.TransactionOuterClass.Transaction;
+import iroha.validation.transactions.storage.TransactionProvider;
+import iroha.validation.util.ObservableRxList;
+import java.security.KeyPair;
+import java.time.Instant;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+import jp.co.soramitsu.iroha.java.BlocksQueryBuilder;
+import jp.co.soramitsu.iroha.java.IrohaAPI;
+import jp.co.soramitsu.iroha.java.Query;
+import jp.co.soramitsu.iroha.java.Utils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class BasicTransactionProvider implements TransactionProvider {
+
+  private static final Logger logger = LoggerFactory.getLogger(MQTransactionProvider.class);
+
+  private final IrohaAPI irohaAPI;
+  private final String accountId;
+  private final KeyPair keyPair;
+  private final Set<String> hashesKnown = new HashSet<>();
+  private final ObservableRxList<Transaction> cache = new ObservableRxList<>();
+
+  public BasicTransactionProvider(IrohaAPI irohaAPI, String accountId, KeyPair keyPair) {
+    Objects.requireNonNull(irohaAPI, "Iroha API must not be null");
+    if (Strings.isNullOrEmpty(accountId)) {
+      throw new IllegalArgumentException("Account ID must not be neither null or empty");
+    }
+    Objects.requireNonNull(keyPair, "Keypair must not be null");
+
+    this.irohaAPI = irohaAPI;
+    this.accountId = accountId;
+    this.keyPair = keyPair;
+    Executors.newScheduledThreadPool(1)
```

Comment:
Running threads from constructors can be risky. In this particular case, I cannot see any problems. But there may be problems in the future. Please take a brief look at [this article](https://www.ibm.com/developerworks/java/library/j-jtp0618/index.html?ca=drs-).

Reply:
Done